### PR TITLE
Pk1d covariance

### PR DESCRIPTION
**Pull request by corentinravoux**
Branch initially designed to compute the covariance matrix during the Pk1D computation.
The computation of this covariance cannot be done directly with simple functions like np.cov() because each sub-forest has a different wavelength binning (both in range and values).
We first need to select the closest wavelength bin for Pk pixels. Then, since we seek to compute the covariance of the one-dimensional power spectrum, the sub-forest need to be sorted by ID. I created a new "chunk_id" for this, which is the concatenation of the forestid and the number of the sub-forest. Then, I computed the covariance matrix term by term. It means that the computation is relatively expensive in terms of calculation time.
Consequently, I parallelized the computation over each (considered-)independent (for now) redshift bin.
I also implemented the bootstrap computation of the covariance, which is "number_bootstrap" times more expensive. I also parallelized this computation over redshift and the number of bootstraps. 
I took this opportunity to parallelize the computation of the average of P1D.
NOTE 1: The covariance matrix is exact, but I do not think this computation will be achievable as it is with a large data set (Y5 for example). We might need to implement an approximate covariance matrix computation.
NOTE 2: The covariance matrix does not (yet) use the SNR weighting scheme newly implemented for average P1D computation.

### CODE REVIEW

**Comment by corentinravoux** [comment]
This branch gives a PyLint error that I do not understand, linked to the delta_extraction part. Maybe @iprafols you know what is happening here?

**Comment by corentinravoux** [comment]
Last commit including the resolving of conflict with master

**Comment by iprafols** [comment]
> This branch gives a PyLint error that I do not understand, linked to the delta_extraction part. Maybe @iprafols you know what is happening here?

They seem to have updated pylint and a few things popped up. I fixed them in another PR so it should be fine now. I checked the pylint output now and there are some things failing for the pk1d side,

**Comment by corentinravoux** [comment]
Ok thanks, now that I merged master I can see it more clearly thanks.

**Review by Waelthus — COMMENTED** [review]
Ok, my main concern is that the computation of covariance and the mean/std quantities can be out-of-sync, e.g. one might choose different weighting schemes for both approaches and get inconsistent results. I'd try to avoid having implementations that can have strongly different results there. I.e. the `fill_average_table` function should have the same definition of what kind of binned average to use that the covar function uses...
I'd also recommend to change both functions to be more easily numbafied, i.e. only use numpy and python, no tables, or i/o routines in there...
Other comments can be found above as well.

**Comment by Waelthus** [comment]
One final question:
Did anyone implement a test for the pk1d_postprocessing and if not did you check if results are the same (in this case a test should be part of this PR)

File: py/picca/pk1d/postproc_pk1d.py [hunk 1]
```diff
@@ -205,229 +201,599 @@ def compute_mean_pk1d(p1d_table,
...
+    )
+    with ThreadPool(number_worker) as pool:
+        pool.starmap(func, params_pool)
+
+    if (compute_covariance) | (compute_bootstrap):
+        if "chunk_id" not in p1d_table.columns:
+            print(
+                """chunk_id not in individual pk files,
+                necessary to compute covariance. Skipping calculation"""
+            )
```
[the remark this answers is not in the document]
Comment (corentinravoux) [reply py/picca/pk1d/postproc_pk1d.py:296]: With the individual pk we have no way to know which sub-forest we are in here I think

File: py/picca/pk1d/postproc_pk1d.py [hunk 2]
```diff
@@ -205,229 +201,599 @@ def compute_mean_pk1d(p1d_table,
...
+    metadata_table["k_max"] = kbin_edges[-1] * np.ones(nbins_z)
+
+    # Initialize cov_table of len = (nzbins * nkbins * nkbins)
+    # corresponding to hdu[3] in final ouput
+    if compute_covariance:
+        cov_table = Table()
+        cov_table["zbin"] = np.zeros(nbins_z * nbins_k * nbins_k)
+        cov_table["index_zbin"] = np.zeros(nbins_z * nbins_k * nbins_k, dtype=int)
+        cov_table["N"] = np.zeros(nbins_z * nbins_k * nbins_k, dtype=int)
+        cov_table["covariance"] = np.zeros(nbins_z * nbins_k * nbins_k)
```
[the remark this answers is not in the document]
Comment (corentinravoux) [reply py/picca/pk1d/postproc_pk1d.py:247]: It did not knew about this error estimate. 
The diagonal error can be computed quickly from the matrix.
What do you mean by the full covariance? you mean the cross-covariance between bootstrap samples?

File: py/picca/pk1d/postproc_pk1d.py [hunk 3]
```diff
@@ -205,229 +201,599 @@ def compute_mean_pk1d(p1d_table,
...
 
     # Initialize mean_p1d_table of len = (nzbins * nkbins) corresponding to hdu[1] in final ouput
     mean_p1d_table = Table()
     nbins_z, nbins_k = len(zbin_edges) - 1, len(kbin_edges) - 1
-    mean_p1d_table['zbin'] = np.zeros(nbins_z * nbins_k)
-    mean_p1d_table['index_zbin'] = np.zeros(nbins_z * nbins_k, dtype=int)
-    mean_p1d_table['N'] = np.zeros(nbins_z * nbins_k, dtype='int64')
+    mean_p1d_table["zbin"] = np.zeros(nbins_z * nbins_k)
+    mean_p1d_table["index_zbin"] = np.zeros(nbins_z * nbins_k, dtype=int)
+    mean_p1d_table["N"] = np.zeros(nbins_z * nbins_k, dtype="int64")
```
[the remark this answers is not in the document]
Comment (corentinravoux) [reply py/picca/pk1d/postproc_pk1d.py:228]: It was used before, I think it change nothing.

File: py/picca/pk1d/postproc_pk1d.py [hunk 4]
```diff
@@ -205,229 +201,599 @@ def compute_mean_pk1d(p1d_table,
...
+            print(
+                """chunk_id not in individual pk files,
+                necessary to compute covariance. Skipping calculation"""
+            )
+            compute_covariance, compute_bootstrap = False, False
+        else:
+            # Initialize cov_table of len = (nzbins * nkbins * nkbins)
+            # corresponding to hdu[3] in final ouput
+            cov_table = Table()
+            cov_table["zbin"] = np.zeros(nbins_z * nbins_k * nbins_k)
```
[the remark this answers is not in the document]
Comment (corentinravoux) [reply py/picca/pk1d/postproc_pk1d.py:302]: It is for the storing in fits FILE, see with Eric for this

File: py/picca/pk1d/postproc_pk1d.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (corentinravoux) [reply py/picca/pk1d/postproc_pk1d.py:296]: yes I agree, chunk_id is a first proposition

File: py/picca/pk1d/postproc_pk1d.py [hunk 5]
```diff
@@ -79,66 +61,74 @@ def read_pk1d(filename, kbin_edges, snrcut=None, zbins_snrcut=None):
...
 
-            chunk_table['forest_z'] = float(chunk_header['MEANZ'])
-            chunk_table['forest_snr'] = float(chunk_header['MEANSNR'])
+            chunk_table["forest_z"] = float(chunk_header["MEANZ"])
+            chunk_table["forest_snr"] = float(chunk_header["MEANSNR"])
+            chunk_table["forest_id"] = int(chunk_header["LOS_ID"])
+            if "CHUNK_ID" in chunk_header:
+                chunk_table[
+                    "chunk_id"
+                ] = f"{chunk_header['LOS_ID']}_{chunk_header['CHUNK_ID']}"
```
[the remark this answers is not in the document]
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py:94]: You mean los_id has not been part of the files or was just ignored, probably the latter. I don't really have a strong opinion regarding using chunk_id to simplify matching, just wanted to note we would have enough info even without.
I guess my intuition that chunks are roughly the same size as zbins failed me here, this is only actually true for z>~2.8, so it should be fine to have this often at lower redshifts.
In principle, you wouldn't need to know anything about the LOS_ID or CHUNK_ID, just if the spectrum is usable and if it falls in the redshift bin of interest as in both, you're averaging over all combinations of chunks anyway later, could be useful for approximations though

File: py/picca/pk1d/postproc_pk1d.py [hunk 6]
```diff
@@ -205,229 +201,599 @@ def compute_mean_pk1d(p1d_table,
...
+        nomedians,
+        nbins_z,
+        zbin_centers,
+        zbin_edges,
+        n_chunks,
+        nbins_k,
+        kbin_edges,
+    )
+    with ThreadPool(number_worker) as pool:
+        pool.starmap(func, params_pool)
```
[the remark this answers is not in the document]
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py:289]: But couldn't it be that a worker reads the full table, tries to write it with its changes and before being finished another worker uses the old copy and changes its things there not knowing about the first processes attempt to do so. Just not sure if changing entries in astropy tables is a thread-safe operation (even if different parts are changed by the workers)

File: py/picca/pk1d/postproc_pk1d.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py:302]: I would do the actual calculation with n-dimensional arrays and only in the end switch to whatever output convention we want. That makes the code more readable and errors less likely by removing all the index computation overheads.

File: py/picca/pk1d/postproc_pk1d.py [hunk 7]
```diff
@@ -205,229 +201,599 @@ def compute_mean_pk1d(p1d_table,
...
+    nomedians,
+    nbins_z,
+    zbin_centers,
+    zbin_edges,
+    n_chunks,
+    nbins_k,
+    kbin_edges,
+    izbin,
+):
+
```
[the remark this answers is not in the document]
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py:430]: Ok, I now realized that one probably needs to run all diagonal elements (incl the means) first, to get the relevant info for the off-diagonal terms. But still I think one would need to have the computations for C_ij and μ_i,σ_i unified

File: py/picca/pk1d/postproc_pk1d.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (corentinravoux) [reply py/picca/pk1d/postproc_pk1d.py:94]: I do not agree with your last statement. If we use all chunk_id in one redshift bin, it would supress the signal as we would compute the cross-correlation between different forest (which might be far also) and we want to compute the covariance for P1D not P3D. 
los_id was previously ignored

File: py/picca/pk1d/postproc_pk1d.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (corentinravoux) [reply py/picca/pk1d/postproc_pk1d.py:228]: done

File: py/picca/pk1d/postproc_pk1d.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (corentinravoux) [reply py/picca/pk1d/postproc_pk1d.py:247]: Do you have the statistical literature for this ? I do not know how we can estimate the error on the covariance matrix, based only on the covariance matrix itself and not on different realizations. Or an additional hypothesis is needed for the statistical distribution of individual pk maybe ?

**Comment by corentinravoux** [comment]
@Waelthus you added too much comment, can you summarize what are the change that we absolutely need for this first version ?

File: py/picca/pk1d/postproc_pk1d.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py:94]: But the cross-correlation between far away forests is something we intrinsically have as part of the P1d measurements as P1d is something different than P3d (kperp~0).

In my view it should be roughly like the following:
- for the bootstrap: you calculate the C_ij = <P_full_set(ki,zi) P_full_set(kj,zj)>_{bootstrap realizations} where indices i,j go over all combinations of k,z that you want to consider. The avg here is over all different bootstrap samples and the P_full_set are computed as the <P> over all chunks contributing to that bin (potentially with some weighting), having chunk_id and los_id simplifies this
- for the sub-sample covariance (with sample size==1): you take all the forests providing information to those bin combinations and compute C_ij = <P_single(ki,zi) P_single(kj,zj)>/(Nforests-1). Also here it would be enough to go through all spectra, test if they contribute to ki, zi, kj, zj and put them into that average. I guess it can be useful to have an index, but it's not required. (Actually I'm not sure anymore how one would do the multi-z correlations this way as there are currently no forests contributing to two different redshift bins, for this purpose one might want to consider either subsamples of size>1 or the different chunks of same spectra falling into different z-bins [which then requires the indexing])
- to speed things up I think one can do the k_binning of the individual forests beforehand which then requires you to store the number of modes going into that allowing to weight each mode the same way

**Comment by Waelthus** [comment]
The very minimal is a test that ensures we're still getting the same results as before (if that does not exist yet). I do think the covariance calculation should at the very least fail if the weighting scheme used for the "mean power" and its std does not match the one for the cov...

File: py/picca/pk1d/postproc_pk1d.py [hunk 8]
```diff
@@ -205,229 +201,599 @@ def compute_mean_pk1d(p1d_table,
...
+            if not np.allclose(delta_z, delta_z[0], atol=1.0e-3):
+                raise ValueError(
+                    "z bins should have equal widths with apply_z_weights."
+                )
+            delta_z = delta_z[0]
+
+            select = (p1d_table["k"] < kbin_edges[ikbin + 1]) & (
+                p1d_table["k"] > kbin_edges[ikbin]
+            )
+            if izbin in (0, nbins_z - 1):
```
[the remark this answers is not in the document]
Comment (armengau) [reply py/picca/pk1d/postproc_pk1d.py:512]: Not sure I understand your question. In general, select is on edges.
At this specific point of the code, there is an exception for z bin selection, when apply_z_weights is True:
in this case, since there is triangle-shape z-weighting ("a la QMLE") you need to extend the selection to nearest bins; on the edges, this leads to a bias (ie for z bin "0" you'll have data from z bin "1" but no data from z bin "-1"...) => there is an exception.
Well, maybe that wasn't your question.

File: py/picca/pk1d/postproc_pk1d.py [hunk 9]
```diff
@@ -205,229 +201,599 @@ def compute_mean_pk1d(p1d_table,
...
+                select = (
+                    select
+                    & (p1d_table["forest_z"] < zbin_centers[izbin + 1])
+                    & (p1d_table["forest_z"] > zbin_centers[izbin - 1])
+                )
 
-            index = (nbins_k * izbin) + ikbin  # index to be filled in table
-            mean_p1d_table['zbin'][index] = zbin_centers[izbin]
-            mean_p1d_table['index_zbin'][index] = izbin
+            redshift_weights = (
```
[the remark this answers is not in the document]
Comment (armengau) [reply py/picca/pk1d/postproc_pk1d.py:527]: Reason = match what's implemented in QMLE

File: py/picca/pk1d/postproc_pk1d.py [hunk 10]
```diff
@@ -205,229 +201,599 @@ def compute_mean_pk1d(p1d_table,
...
+
+    if n_chunks[izbin] == 0:  # Fill rows with NaNs
+        zbin_array[:] = zbin_centers[izbin]
+        index_zbin_array[:] = izbin
+        n_array[:] = 0
+        covariance_array[:] = np.nan
+        return zbin_array, index_zbin_array, n_array, covariance_array
+
+    for chunkid in chunkids:  # First loop 1) id forest bins
+        select_id = select_z & (p1d_table["chunk_id"] == chunkid)
```
[the remark this answers is not in the document]
Comment (armengau) [reply py/picca/pk1d/postproc_pk1d.py:733]: Same question!

File: py/picca/pk1d/postproc_pk1d.py [hunk 11]
```diff
@@ -450,27 +816,45 @@ def run_postproc_pk1d(data_dir,
...
+
+    result = fitsio.FITS(output_file, "rw", clobber=True)
+    result.write(mean_p1d_table.as_array())
+    result.write(
+        metadata_table.as_array(),
+        header={"VELUNITS": velunits, "NQSO": len(np.unique(p1d_table["forest_id"]))},
+    )
+    if cov_table is not None:
+        result.write(cov_table.as_array())
     result.close()
```
[the remark this answers is not in the document]
Comment (armengau) [reply py/picca/pk1d/postproc_pk1d.py:860]: @Waelthus I would naturally have stored the covmat as an image HDU (more precisely a 3D array), indeed with covmat.shape = (nz, nk, nk). Would it be ok to do that?

[the remark this answers is not in the document]
Comment (armengau) [reply py/picca/pk1d/postproc_pk1d.py:860]: About your point on Table.write: I think I tried this and then had issue with the custom header (does the option "meta" do the same as FITS's header?). Can be tested again.

**Comment by armengau** [comment]
Summary of minor items from our zoom:
- ~l. 94  chunck_table['chunk_id'] = (los_id)_(chunk_id) => change name
- add doc/warning/errors if both options covariance & (z_weight / SNRweight)
- covmat_table has duplicate initialisations (l.240 & l.301)
- fct name "fill_average_table" => include "zbin" in the name ?

File: py/picca/pk1d/postproc_pk1d.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py:247]: It's coming from the covariance matrix being drawn from a Wishart distribution. That might help finding more info...

File: py/picca/pk1d/postproc_pk1d.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py:860]: in principle it should do write the dict inside meta into the header, not that I'm against the current implementation. Just wanted to note it could be simpler.

Regarding storage: I would do as you say store it in a (nz, nk, nk) structure as long as we're not considering nk being dependent on z, if we want to implement that (and we might need to) then it indeed makes more sense to do a nz*nk(z)*nk(z) structure, but also the whole algorithm would become more complicated

File: py/picca/pk1d/postproc_pk1d.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py:512]: ah, I see. So the selection in the weighted case  is such that for bin i you'd select data from the center of bin i-1 to the center of bin i+1, in total spanning 2 full bins around the center of bin i, with a weighting scheme downweighting far away points.
In that case this should be completely fine, just didn't properly read what is going on I guess...
How is the weighting done for the corner cases? Do we get a bias there from the asymmetrically distributed data wrt the bin center...

[the remark this answers is not in the document]
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py:512]: ah, got it. So the boundary bins only take the data from within that bin, but elsewise apply the same weighting scheme as other bins which additionally take half a bin outside in each direction into account. Which removes the asymmetry and therefore also the biasing.
Thanks for those explanations!

File: py/picca/pk1d/postproc_pk1d.py [hunk 12]
```diff
@@ -205,229 +201,599 @@ def compute_mean_pk1d(p1d_table,
...
+        )
+        with Pool(number_worker) as pool:
+            output_cov = pool.starmap(func, params_pool)
+
+        for izbin in range(nbins_z):  # Main loop 1) z bins
+            zbin_array, index_zbin_array, n_array, covariance_array = (
+                *output_cov[izbin],
+            )
+            i_min = izbin * nbins_k * nbins_k
+            i_max = (izbin + 1) * nbins_k * nbins_k
```
[the remark this answers is not in the document]
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py:349]: I'm fine with those either way in the end as I now realized that changing nbins_k to nbins_k(z) will require this step anyway (just with some additional complications) at some point

File: py/picca/pk1d/postproc_pk1d.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py:302]: this is fine either way, I see that there might be a gain doing it like this already when moving to nbins_k(z) later.

File: py/picca/pk1d/postproc_pk1d.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py]: e.g. discussed in here (around eq 18): https://arxiv.org/pdf/1212.4359.pdf

**Comment by corentinravoux** [comment]
Merging master for testing the branch on data.

**Comment by corentinravoux** [comment]
Test on EDR+M2 data: the correlation matrix gives similar trend than eBOSS.

test_covariance_pk1d

**Comment by Waelthus** [comment]
The red block at large scales is curious. 1) For eBOSS correlations at those scales (roughly the first 10 bins in the P(k) measurement) seem to be less strong, 2) this is roughly the region where eBOSS and DESI-EDR+M2 disagree. So naively one would overestimate the disagreement, which I'd take as a reason to at least show the covar estimate somewhere in the paper (but do pcolormesh with a proper x/y-axis in that case), at least in an Appendix...
Is this similar for all the redshifts?

**Comment by corentinravoux** [comment]
The P1D EDR+M2 is already too long and we cannot compare directly this matrix result with the eBOSS one which is smoothed. More code is needed to include a covariance matrix in a paper.

**Comment by corentinravoux** [comment]
I am going in long vacation. I suggest to push this PR since it gives the same result than previously for the average P1D and is able to compute properly a covariance matrix.
All other modifications (covariance matrix improvement, p1d testing, ...) should be added on another PR.

**Comment by Waelthus** [comment]
Regarding the merge. Given that we can at least still produce results from master things would be OK with me. Especially if your away for a longer time.
Re the correlation: I don't think that a correlation estimate for at least one redshift bin without any smoothing (ie the plot shown above or is this an average over all z) potentially in an appendix would hurt. Any smoothing can be done after if someone wants to use the measurement. With the correlations we're seeing here, the basically couldn't be used with just diagonal errorbars so we might wanna add a disclaimer (if we didn't already, forgot if we did). Also: how strong does the correlation become at small k? Without the truncation at 0.25?

**Comment by corentinravoux** [comment]
I am back from vaccation. We should discuss the state of this branch at the next teleconf. For me, it is a first version which works but could be improved.

**Comment by iprafols** [comment]
I'd say to merge now and do another smaller PR, this way it'll be easier to review

File: py/picca/pk1d/postproc_pk1d.py [hunk 13]
```diff
@@ -205,229 +200,791 @@ def compute_mean_pk1d(p1d_table,
     One row per z bin; column values z_min/max, k_min/max, N_chunks
     """
```
Comment (iprafols) [inline py/picca/pk1d/postproc_pk1d.py:201]: docstring needs fixing

File: py/picca/pk1d/postproc_pk1d.py [hunk 14]
```diff
@@ -205,229 +200,791 @@ def compute_mean_pk1d(p1d_table,
...
-    metadata_table['z_min'] = zbin_edges[:-1]
-    metadata_table['z_max'] = zbin_edges[1:]
-    metadata_table['k_min'] = kbin_edges[0] * np.ones(nbins_z)
-    metadata_table['k_max'] = kbin_edges[-1] * np.ones(nbins_z)
+    metadata_table["z_min"] = zbin_edges[:-1]
+    metadata_table["z_max"] = zbin_edges[1:]
+    metadata_table["k_min"] = kbin_edges[0] * np.ones(nbins_z)
+    metadata_table["k_max"] = kbin_edges[-1] * np.ones(nbins_z)
+
+    if (compute_covariance) | (compute_bootstrap):
```
Comment (iprafols) [inline py/picca/pk1d/postproc_pk1d.py:239]: use `or` unless you are doing a bitwise OR (it's not the case here, is it?) Also you can drop the unnecessary parenthesis

**Review by iprafols — CHANGES_REQUESTED** [review]
I found some minor things that I think should be fixed before merging. Also, I noted that there are several unresolved comments/questions from @Waelthus. It's fine if some are left for a latter time, but we should make sure @Waelthus is happy with going ahead before merging

**Comment by iprafols** [comment]
ok so then let's merge once my two minor things  are fixed

File: py/picca/pk1d/postproc_pk1d.py (hunk 14 above)
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py]: I fixed this one

File: py/picca/pk1d/postproc_pk1d.py (hunk 13 above)
Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py:201]: not sure what you mean here, fixed the docstrings to be numpy-docstring compatible according to pydocstyle (maybe this should be run at linting time as well, to have properly formatted docstrings), the line cited is just regarding the format the table needs, which I think is described reasonably well...

**Comment by Waelthus** [comment]
@iprafols If you're fine with my changes above, we can merge now.

File: py/picca/pk1d/postproc_pk1d.py (hunk 13 above)
Comment (iprafols) [reply py/picca/pk1d/postproc_pk1d.py:201]: I mean that there are several keyword arguments that are not described on the docstring. Basically all these
```
nomedians=False,
    velunits=False,
    output_snrfit=None,
    compute_covariance=False,
    compute_bootstrap=False,
    number_bootstrap=50,
    number_worker=8
```

Comment (Waelthus) [reply py/picca/pk1d/postproc_pk1d.py:201]: Ah, I'd say we can merge with an incomplete docstring given that we want to update after anyway. And most of those variable names are self-explanatory anyway... But if you insist, I guess one could add those

Comment (armengau) [reply py/picca/pk1d/postproc_pk1d.py:215]: I've completed the docstring  on covariance options (the rest is documented in either of the functions).

**Review by iprafols — APPROVED** [review]
